### ISSUE TITLE
Design Doc: Complete Fluid

### ISSUE THREAD

**Issue by wangkuiyi** [PaddlePaddle/Paddle#10103]
## Motivation

Currently, our Fluid implementation is incomplete and makes the AI application programming pretty nasty.

- Fluid was designed to make AI programming concise and easy, like examples in
   - https://github.com/PaddlePaddle/Paddle/issues/7464
   - https://github.com/PaddlePaddle/Paddle/issues/9912

- However, the currently runnable examples are as follows
  - [`test_fit_a_line.py`](https://github.com/PaddlePaddle/Paddle/blob/bcf7c36b0b3d62caeea351d9905ac901cb7a1f26/python/paddle/fluid/tests/book/test_fit_a_line.py)

In order to complete the design of Fluid, we create this issue as a table of context of the design of Fluid.

## Questions to Answer

1. How to train on one node
   1. Read data: https://github.com/PaddlePaddle/Paddle/issues/10102
      1. Training data augmentation: augmentation programs *pipes* the output to PaddlePaddle programs
      1. The data format: RecordIO is good for training/testing data, but not good for model parameters. https://github.com/wangkuiyi/recordio
      1. Map records into Fluid variables.
   1. Save model parameters
   1. How to write the program https://github.com/PaddlePaddle/Paddle/issues/9912
   1. How to fetch and feed data from the front end language
   1. Should we allow the ability to fetch intermediate variables during execution?
   1. How to run the program
   1. How to debug
   1. Do we want to allow running arbitrary Python code during training? E.g., some custom Python logic to control save model or not during training.
   1. How to fetch data during training steps In discussion: https://github.com/PaddlePaddle/Paddle/issues/10152
   1. How to initialize a Fluid program? In discussion: https://github.com/PaddlePaddle/Paddle/issues/10177 

1. How to do distributed training
   1. [How to do distributed reading](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/fluid/design/dist_train/distributed_architecture.md)
   1. How to save huge models
   1. [How to write program](https://github.com/PaddlePaddle/Paddle/tree/develop/doc/fluid/design/dist_train)
   1. How to run the program
      - SLURM + container
      - MPI + container
      - [Kubernetes](https://github.com/PaddlePaddle/edl)

1. How to do inference
   1. Easy experiments
   1. Large-scale serving
   1. How to export to ONNX and pipe with TensorRT

### Examples of [an API proposal](https://github.com/PaddlePaddle/Paddle/issues/10152)

- [Recognize Digits](https://github.com/PaddlePaddle/Paddle/issues/10215)
- [Label Semantic Roles](https://github.com/PaddlePaddle/Paddle/issues/10216)
- [Word to Vec](https://github.com/PaddlePaddle/Paddle/issues/10214)

**Comment by helinwang** [comment]
Thanks @JiayiFeng ! The new design will change some API to make Fluid more user friendly, but the underlying data format will still be RecordIO. If the current cpp reader is not compatible with the new API, we will re-use the current cpp reader implementation as much as possible.

**Comment by shanyi15** [comment]
您好，此issue在近一个月内暂无更新，我们将于今天内关闭。若在关闭后您仍需跟进提问，可重新开启此问题，我们将在24小时内回复您。因关闭带来的不便我们深表歉意，请您谅解~感谢您对PaddlePaddle的支持! 
 Hello, this issue has not been updated in the past month. We will close it today for the sake of other user‘s experience. If you still need to follow up on this question after closing, please feel free to reopen it. In that case, we will get back to you within 24 hours. We apologize for the inconvenience caused by the closure and thank you so much for your support of PaddlePaddle Group!
